integration/airflow/openlineage/airflow/extractors/sftp_extractor.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2018-2022 contributors to the OpenLineage project
+# SPDX-License-Identifier: Apache-2.0
+import socket
+from typing import Optional, List
+
+from openlineage.airflow.extractors.base import BaseExtractor, TaskMetadata
+from openlineage.common.dataset import Source, Dataset
+from airflow.providers.sftp.operators.sftp import SFTPOperation
+from airflow.providers.ssh.hooks.ssh import SSHHook
+
+
+class SFTPExtractor(BaseExtractor):
+    def __init__(self, operator):
+        super().__init__(operator)
+
+    @classmethod
+    def get_operator_classnames(cls) -> List[str]:
+        return ["SFTPOperator"]
+
+    def extract(self) -> Optional[TaskMetadata]:
+        scheme = "file"
+        local_host = socket.gethostname()
+
+        if hasattr(self.operator, "remote_host") and self.operator.remote_host is not None:
+            remote_host = self.operator.remote_host
+        elif hasattr(self.operator, "sftp_hook") and self.operator.sftp_hook is not None:
+            remote_host = self.operator.sftp_hook.get_connection().host
+        elif hasattr(self.operator, "ssh_hook") and self.operator.ssh_hook is not None:
```

Comment:
> SFTPOperator always has remote_host, sftp_hook, ssh_hook and I don't think this is needed to use hasattr in each case.

I don't think that it's always true, for example in case that only `ssh_conn_id` is provided and neither `sftp_hook` nor `ssh_hook` is specified by user. In such case, [both hook variables are `None` within `SFTPOperator`'s constructor](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L129-L146) and then [`sftp_hook` is set in its `execute` method](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L151-L158). So if we try to get `remote_host` from `sftp_hook` or `ssh_hook` in `SFTPExtractor.extract()`, those attributes are `None` and `AttributeError` is raised.

I also tried to use `BaseExtractor`'s `extract_on_complete` method instead of `extract` so as to access `sftp_hook` set during `SFTPOperator.execute()` via the `task_instance` parameter, but its `task` attribute doesn't seem to be set always, as commented [here](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/models/taskinstance.py#L517). `self.operator` also seems to keep the default state even in `extract_on_complete`, so I chose the current implementation.

> Also there's a logic in operator that sets sftp_hook with SFTPHook(ssh_hook=self.ssh_hook).

That logic is executed only when `ssh_hook` is explicitly provided by user.
https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L134-L146

Reply:
> I don't think that it's always true, for example in case that only ssh_conn_id is provided and neither sftp_hook nor ssh_hook is specified by user. In such case, [both hook variables are None within SFTPOperator's constructor](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L129-L146) and then [sftp_hook is set in its execute method](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L151-L158). So if we try to get remote_host from sftp_hook or ssh_hook in SFTPExtractor.extract(), those attributes are None and AttributeError is raised.

You're right. I'm sorry for pointing it out incorrectly. That's still quite odd that `sftp_hook` might be set in two different places.

> I also tried to use BaseExtractor's extract_on_complete method instead of extract so as to access sftp_hook set during SFTPOperator.execute() via the task_instance parameter, but its task attribute doesn't seem to be set always, as commented [here](https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/models/taskinstance.py#L517). self.operator also seems to keep the default state even in extract_on_complete, so I chose the current implementation.

+1

>>Also there's a logic in operator that sets sftp_hook with SFTPHook(ssh_hook=self.ssh_hook).
>
> That logic is executed only when ssh_hook is explicitly provided by user.
> https://github.com/apache/airflow/blob/providers-sftp/4.1.0/airflow/providers/sftp/operators/sftp.py#L134-L146

I meant here to skip checking if `ssh_hook` is set as `sftp_hook` overtakes it (`SFTPHook` accepts `ssh_hook` argument - it's marked deprecated anyway). However, we could assume we don't know how operator's code will change. Let's keep it as it is now just to be safe.